### DISCUSSION TITLE
any pure erlang example

### DISCUSSION THREAD

**Discussion by benoitc** [GetFirefly/firefly#375] [Q&A]
Do you have any example of its usage using Erlang (and not elixir) that are not imply npm also?

**Comment by KronicDeth** [comment]
The fib.erl example for the interpreter.

**Comment by KronicDeth** [comment]
Expanding on my last answer.

Following the [`liblumen_eir_interpreter` README](https://github.com/lumen/lumen/tree/develop/liblumen_eir_interpreter#elixir-code), but skipping the decompile step because we're not using Elixir:

## For `liblumen_eir_interpreter/examples/fib`

1. `cd liblumen_eir_interpreter`
2. `cargo run -- --ident fib:run/0 examples/fib/fib.erl`

## For `liblumen_eir_interpreter/examples/simple_server`

1. `cd liblumen_eir_interpreter`
2. `cargo run -- --ident foo:run/0 examples/simple_server/*`

Are those what you were looking for @benoitc?

**Comment by benoitc** [comment]
yes defintely. this is interesting i will try it and report there.

**Comment by KronicDeth** [comment]
@benoitc The `lumen` tests ([hello_world.rs](https://github.com/lumen/lumen/blob/8cfd9dfeeecca1c29da9fab3a4feafeeae4d9624/lumen/tests/hello_world.rs) and [cli.rs](https://github.com/lumen/lumen/blob/develop/lumen/tests/cli.rs)) use Erlang ([hello_world/init.erl](https://github.com/lumen/lumen/blob/8cfd9dfeeecca1c29da9fab3a4feafeeae4d9624/lumen/tests/hello_world/init.erl) and [cli/init.erl](https://github.com/lumen/lumen/blob/develop/lumen/tests/cli/init.erl), respectively) with the compile (`lumen`).  The one restriction with the compiler currently is that the entry point must be `init:start/0`, so it need to be exported from your own code:

## `hello_world`

https://github.com/lumen/lumen/blob/8cfd9dfeeecca1c29da9fab3a4feafeeae4d9624/lumen/tests/hello_world/init.erl#L1-L3

https://github.com/lumen/lumen/blob/8cfd9dfeeecca1c29da9fab3a4feafeeae4d9624/lumen/tests/hello_world/init.erl#L8-L10


## `cli`

https://github.com/lumen/lumen/blob/8cfd9dfeeecca1c29da9fab3a4feafeeae4d9624/lumen/tests/cli/init.erl#L1-L2
https://github.com/lumen/lumen/blob/8cfd9dfeeecca1c29da9fab3a4feafeeae4d9624/lumen/tests/cli/init.erl#L5-L15

**Comment by KronicDeth** [comment]
All the integration tests for `liblumen_otp` are now written as the Rust checking the stdout of `lumen` compiled `.erl` files.  All of those can be found under https://github.com/lumen/lumen/tree/develop/native_implemented/otp/tests.

**Comment by KronicDeth** [comment]
You can also use the language search to get [all the Erlang files in the repo in general](https://github.com/lumen/lumen/search?l=erlang).  This can be done from the UI by clicking on Erlang in the Language composition on the project home:

1. Go to https://github.com/lumen/lumen
2. Click Erlang in this widget
   ![Screen Shot 2020-08-12 at 9 28 37 AM](https://user-images.githubusercontent.com/298259/90027495-37e58480-dc7e-11ea-9b40-e6366c826aff.png)

**Comment by 221V** [comment]
> Expanding on my last answer.
> 
> Following the [`liblumen_eir_interpreter` README](https://github.com/lumen/lumen/tree/develop/liblumen_eir_interpreter#elixir-code), but skipping the decompile step because we're not using Elixir:
> ## For `liblumen_eir_interpreter/examples/fib`

this link has "404 not found" status..

but thank you for examples in messages under this one!
thank you the much!

really! erlang-in-wasm = the great idea!
thank you for this implementation! :+1: :))

**Comment by 221V** [comment]
also one more question --

can i try to prog a tiny io game with lumen for wasm right now?
is lumen "just for fun coding"-ready? or not yet? 

not for any production -- just for fun only -- coding a bit + play with friend a bit!! ))

**Comment by KronicDeth** [comment]
@221V The `lumen` compiler we just release cannot target WASM yet (we're trying to nudge the WASM Community Group along to add stack switching to make WASM more like x8_64), so when targeting WASM you can have the runtime and OTP BIFs, but the layer that is your application needs to either be ported to Rust like I did for the spawn-chain WASM demo OR use the browser interpreter, but IDK if the interpreter is in a working state.  We had to break it to get x86_64 compiling working.  If you're interested in using the interpreter, then @hansihe  can help you.  If you're interested in porting to Rust for the application layer, I can help you.

**Comment by 221V** [comment]
@KronicDeth i do not know Rust yet :)
and only a bit can read C code (not C++)
and i do not know ok FASM - just newbie-beginner, noob)

i know ok (at this moment) --
- vanilla.js in browser only (no OOP - procedural only),

- php (like js - no OOP, but i was start from it - 12 years ago, and few last years not working with this),

- erlang (only one moment not ok yet --
  i have problem to read fully-parallel-code -
  when minimal FP code, max actors-everywhere)

- haskell (not "hard things" - mainly not that high-math-root things,
  "easy way" only now - but my haskell code ok nice works, i am happy)

so FP is omw -- i have planning became true-senior with erlang and haskell now,
and try get the same with Rust and FASM in future :)

and no more langs, no more other --
cause there are many other interesting things in our world ))
not programming and pc/internet gaming-and-programming only))

i think so)

thank you! :)
thanks for true-anwer, cool answer!

there are many cool wasm examples-demos in the internet -
but.. i can not read and use that all))
i do not know OOP js,
i do not know C++ (and want not lose time for this two),
and i do not know the Rust yet ) cause needs more free time (and/or a  bit more money in family) for learn it :)

and examples Haskell -> wasm = oh.. "hard things" again (high-math-root things),
 + +not fully wasm yet, yes? :)

so i am in search for get cool tool-for-deal-with-wasm in browsers :)
thats why i ask question(s) here (:

i see lumen is the cool great idea!
really cool for me! <3 and need tool like it! ))


thank you a lot! :)
